File: MidtermLabWork1/JAVA/ProfitabilityAnalysis.java
package MidtermLabWork1.JAVA;

import java.io.*;
import java.util.*;

public class ProfitabilityAnalysis {

    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        File file;
        String path;

        // Step 1: Prompt and Validate File Path
        while (true) {
            System.out.print("Enter dataset file path (e.g., C:\\Downloads\\vgchartz-2024.csv): ");
            path = input.nextLine();
            file = new File(path);

            if (file.exists() && file.isFile() && file.canRead()) {
                if (path.toLowerCase().endsWith(".csv")) {
                    break;
                } else {
                    System.out.println("Error: File is not in CSV format.");
                }
            } else {
                System.out.println("Error: Invalid file path or file is not readable. Please try again.");
            }
        }

        // Step 2: Load and Process Data
        List<DataRecord> records = loadDataset(path);

        if (records.isEmpty()) {
            System.out.println("No data found or error reading the file.");
            input.close();
            return;
        }

        // Step 3: Perform Analytics
        performAnalytics(records);
        
        input.close();
    }

    private static List<DataRecord> loadDataset(String filePath) {
        List<DataRecord> records = new ArrayList<>();
        String line;

        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            // Skip header
            br.readLine();
            
            while ((line = br.readLine()) != null) {
                // Using regex split to handle potential commas inside quotes (CSV standard)
                String[] values = line.split(",(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)");
                
                // Columns: 1=title, 3=genre (category), 7=total_sales
                if (values.length > 7) {
                    try {
                        String title = values[1];
                        String category = values[3];
                        String salesStr = values[7].trim();
                        
                        if (!salesStr.isEmpty() && !salesStr.equalsIgnoreCase("N/A")) {
                            double sales = Double.parseDouble(salesStr);
                            records.add(new DataRecord(title, category, sales));
                        }
                    } catch (NumberFormatException e) {
                        // Skip rows with invalid numeric sales values
                        continue;
                    }
                }
            }
        } catch (IOException e) {
            System.out.println("Critical Error reading file: " + e.getMessage());
        }
        return records;
    }

    private static void performAnalytics(List<DataRecord> records) {
        Map<String, Double> totalSalesMap = new HashMap<>();
        Map<String, Integer> countMap = new HashMap<>();

        // Grouping data
        for (DataRecord record : records) {
            String cat = record.getCategory();
            double sales = record.getSales();

            totalSalesMap.put(cat, totalSalesMap.getOrDefault(cat, 0.0) + sales);
            countMap.put(cat, countMap.getOrDefault(cat, 0) + 1);
        }

        // Analytics variables
        String mostProfitableCat = "";
        double maxSales = -1;
        String leastProfitableCat = "";
        double minSales = Double.MAX_VALUE;

        // Formatted Output Header
        System.out.println("\n" + "=".repeat(60));
        System.out.printf("%-20s | %-15s | %-15s\n", "Category", "Total Sales", "Avg Sale");
        System.out.println("-".repeat(60));

        for (String category : totalSalesMap.keySet()) {
            double totalSales = totalSalesMap.get(category);
            int count = countMap.get(category);
            double avgSale = totalSales / count;

            System.out.printf("%-20s | $%-14.2f | $%-14.2f\n", category, totalSales, avgSale);

            // Check for most/least profitable
            if (totalSales > maxSales) {
                maxSales = totalSales;
                mostProfitableCat = category;
            }
            if (totalSales < minSales) {
                minSales = totalSales;
                leastProfitableCat = category;
            }
        }

        // Display results
        System.out.println("=".repeat(60));
        System.out.println("ANALYTICS SUMMARY:");
        System.out.printf("Most Profitable Category:  %s ($%.2f)\n", mostProfitableCat, maxSales);
        System.out.printf("Least Profitable Category: %s ($%.2f)\n", leastProfitableCat, minSales);
        System.out.println("=".repeat(60));
    }
}

File: MidtermLabWork1/JAVASCRIPT/ProfitabilityAnalysis.js
/**
 * ProfitabilityAnalysis.js
 * Performs product category profitability analysis
 */

class ProfitabilityAnalysis {
    /**
     * Analyze profitability by category
     */
    static analyzeByCatego(records) {
        const categoryData = {};

        // Group records by category (genre)
        records.forEach(record => {
            const category = record.genre || 'Unknown';
            
            if (!categoryData[category]) {
                categoryData[category] = {
                    category: category,
                    records: [],
                    totalSales: 0,
                    count: 0
                };
            }

            const sales = parseFloat(record.totalSales) || 0;
            categoryData[category].records.push(record);
            categoryData[category].totalSales += sales;
            categoryData[category].count += 1;
        });

        // Calculate averages and format results
        const results = Object.values(categoryData).map(cat => ({
            category: cat.category,
            totalSales: parseFloat(cat.totalSales.toFixed(2)),
            averageSales: parseFloat((cat.totalSales / cat.count).toFixed(2)),
            productCount: cat.count
        }));

        // Sort by total sales descending
        results.sort((a, b) => b.totalSales - a.totalSales);

        return results;
    }

    /**
     * Get profitability statistics
     */
    static getStatistics(results) {
        if (results.length === 0) {
            return null;
        }

        return {
            mostProfitable: results[0],
            leastProfitable: results[results.length - 1],
            categories: results
        };
    }

    /**
     * Display analysis results in formatted table
     */
    static displayResults(statistics) {
        if (!statistics) {
            console.log('No data available for analysis.');
            return;
        }

        console.log('\n' + '='.repeat(100));
        console.log('PRODUCT CATEGORY PROFITABILITY ANALYSIS'.padEnd(100));
        console.log('='.repeat(100));

        console.log('\n┌─ CATEGORY SUMMARY ─────────────────────────────────────────────────────────────────────┐');
        console.log('│ Rank │ Category              │ Total Sales (M) │ Avg Sales (M) │ # Products │');
        console.log('├──────┼───────────────────────┼─────────────────┼───────────────┼────────────┤');

        statistics.categories.forEach((cat, index) => {
            const rank = String(index + 1).padEnd(5);
            const category = cat.category.substring(0, 21).padEnd(23);
            const totalSales = String(cat.totalSales).padStart(15);
            const avgSales = String(cat.averageSales).padStart(13);
            const count = String(cat.productCount).padStart(10);
            
            console.log(`│ ${rank}│ ${category}│ ${totalSales} │ ${avgSales} │ ${count} │`);
        });
        console.log('└──────┴───────────────────────┴─────────────────┴───────────────┴────────────┘');

        console.log('\n┌─ KEY INSIGHTS ─────────────────────────────────────────────────────────────────────────┐');
        console.log(`│ Most Profitable Category:   ${statistics.mostProfitable.category.padEnd(70)} │`);
        console.log(`│   • Total Sales: ${String(statistics.mostProfitable.totalSales + 'M').padEnd(70)} │`);
        console.log(`│   • Average per Product: ${String(statistics.mostProfitable.averageSales + 'M').padEnd(70)} │`);
        console.log(`│   • Product Count: ${String(statistics.mostProfitable.productCount).padEnd(70)} │`);
        
        console.log(`│                                                                                      │`);
        console.log(`│ Least Profitable Category:  ${statistics.leastProfitable.category.padEnd(70)} │`);
        console.log(`│   • Total Sales: ${String(statistics.leastProfitable.totalSales + 'M').padEnd(70)} │`);
        console.log(`│   • Average per Product: ${String(statistics.leastProfitable.averageSales + 'M').padEnd(70)} │`);
        console.log(`│   • Product Count: ${String(statistics.leastProfitable.productCount).padEnd(70)} │`);
        console.log('└────────────────────────────────────────────────────────────────────────────────────────┘');

        console.log('\n' + '='.repeat(100) + '\n');
    }
}

module.exports = ProfitabilityAnalysis;
